### PR TITLE
QuantityMean

### LINKED ISSUE #143: level means estimates
Always pass level means and vars to the MeanQuantity, possibly compute the total mean and var on request. That means a richer API of MeanQuantity.

### LINKED ISSUE #145: QuantityMean._reshape
`QuantityMean._reshape()`

What is `data`? Rename to something more specific and provide a comment or type hint.
What is the meaning of the`levels` flag? It seems to me that there should possibly be two methods, one calling the other.

### LINKED ISSUE #146: QuantityMean.__getitem__
`QuantityMean.__getitem__`

Try to simplify this method:
1.  do not store mean and var, only l_mean and l_var
2. reduce number of special cases, keep QuantityMean in a single consistent structure. e.g. neither reshape_shape nor newshape should be None (fill them with proper values before).
3. If you find the special cases inevitable put there a comment to explain what are the individual cases.

Possibly also try to break the method into two parts:

    

1. convert general key argument to a particular slice ... via QType
2. extract part of l_means and l_vars (there should be a method for that to make the same extraction on both) ... independent of QType

### PR DESCRIPTION
**Pull request by martinspetlik**
resolve #143
resolve #145
resolve #146

### CODE REVIEW

**Comment by martinspetlik** [comment]
resolve #143 
resolve #145 
resolve #146

File: src/mlmc/quantity.py [hunk 1]
```diff
@@ -538,126 +538,113 @@ def samples(self, chunk_spec):
...
         """
-        Return reshaped mean
+        Calculates the overall estimates of the mean and the variance from the means and variances at each level
         """
-        return self.mean
+        self._mean = np.sum(self._l_means, axis=0)
+        self._var = np.sum(self._l_vars / self._n_samples[:, None], axis=0)
+
+        # Sums full of zeros leads to NaN
+        if np.isnan(self._mean).all():
```
Comment (jbrezmorf) [inline src/mlmc/quantity.py:567]: Why this special case?

File: src/mlmc/quantity.py [hunk 2]
```diff
@@ -538,126 +538,113 @@ def samples(self, chunk_spec):
 
 class QuantityMean:
 
-    def __init__(self, quantity_type, mean, var, l_means=[], l_vars=[], n_samples=None, n_rm_samples=0):
+    def __init__(self, quantity_type, l_means, l_vars, n_samples=None, n_rm_samples=None):
```
Comment (jbrezmorf) [inline src/mlmc/quantity.py]: Make n_samples and n_rm_samples obligatory. None values are not valid.

File: src/mlmc/quantity.py (hunk 1 above)
Comment (jbrezmorf) [reply src/mlmc/quantity.py:567]: Rather split this method into mean and var property functions.

File: src/mlmc/quantity.py [hunk 3]
```diff
@@ -538,126 +538,113 @@ def samples(self, chunk_spec):
...
         """
-        Get items from Quantity, quantity type must support brackets access
+        Get item from current QuantityMean, quantity type must support brackets access
+        All levels means and vars are reshaped to their QType shape and then the item is gotten,
+        ath the end, new QuantityMean instance is created with flatten selected means and vars
         :param key: str, int, tuple
-        :return: np.ndarray
+        :return: QuantityMean
         """
         new_qtype, start = self.qtype[key]  # New quantity type
```
Comment (jbrezmorf) [inline src/mlmc/quantity.py]: Seems, that the current behavior of qtype[key] violates the natural semantics of the indexing as it returns 'start' as well. 
In such a case we should rather use a regular method.

File: src/mlmc/quantity.py [hunk 4]
```diff
@@ -538,126 +538,113 @@ def samples(self, chunk_spec):
...
+            new_shape = l_means.shape[0], *reshape_shape, np.prod(l_means.shape[1:]) // np.prod(reshape_shape)
+            l_means = l_means.reshape(new_shape)
+            l_vars = l_vars.reshape(new_shape)
+
+        # Getting items
+        l_means = l_means[:, key]
+        l_vars = l_vars[:, key]
 
+        # Flatten l_means and l_vars, keep the first item of the shape equal to the number of levels
         if len(l_means.shape) > 1:
```
Comment (jbrezmorf) [inline src/mlmc/quantity.py]: Preform always.

File: src/mlmc/quantity.py [hunk 5]
```diff
@@ -538,126 +538,113 @@ def samples(self, chunk_spec):
...
         :param key: str, int, tuple
-        :return: np.ndarray
+        :return: QuantityMean
         """
         new_qtype, start = self.qtype[key]  # New quantity type
         reshape_shape = None
-        newshape = None
+
         # ArrayType might be accessed directly regardless of qtype start and size
         if isinstance(self.qtype, qt.ArrayType):
```
Comment (jbrezmorf) [inline src/mlmc/quantity.py]: getitem should be common to the Quantity and thus implemented in QType as it is the part shared between Quantity and QuantityMean. It should go as:
```
new_qtype, slice = self.qtype.get_key(key)

samples_like = data[None, None,:]     # applied for data in [self._l_means, self._l_vars]
shaped_data = self.qtype.reshape(samples_like)     # actually reshape only for Array type
new_data = shaped_data[slice].ravel()
```

File: src/mlmc/quantity_estimate.py [hunk 6]
```diff
@@ -68,31 +70,16 @@ def estimate_mean(quantity, chunk_size=512000000, level_means=False):
...
-            if n > 1:
-                l_vars.append((sp - (s ** 2 / n)) / (n-1))
-            else:
-                l_vars.append((sp - (s ** 2)))
-
-    # sums full of zeros
-    if np.isnan(mean).all():
-        mean = []
-        var = []
+            l_vars.append((sp - (s ** 2)))
```
Comment (jbrezmorf) [inline src/mlmc/quantity_estimate.py]: for n<=1 the variance should be `inf`.

File: src/mlmc/quantity_types.py [hunk 7]
```diff
@@ -75,6 +75,9 @@ def _make_getitem_op(self, chunk, key):
         """
         return QType.keep_dims(chunk[key])
```
Comment (jbrezmorf) [inline src/mlmc/quantity_types.py:76]: Considering the file name, we should possibly introduce all classes into `mlmc` module via. `mlmc.__init__.py`.

File: src/mlmc/quantity_types.py [hunk 8]
```diff
@@ -136,6 +143,12 @@ def _make_getitem_op(self, chunk, key):
         chunk = chunk.reshape((*self._shape, chunk.shape[-2], chunk.shape[-1]))
         return QType.keep_dims(chunk[key])
 
+    def reshape(self, data):
+        if isinstance(self._qtype, ScalarType):
+            return data.reshape(self._shape)
```
Comment (jbrezmorf) [inline src/mlmc/quantity_types.py:148]: Reshape consistently to (*self._shape, 1).

Comment (jbrezmorf) [reply src/mlmc/quantity_types.py:148]: Is it compatible with reshaping the samples data?

File: src/mlmc/quantity.py (hunk 4 above)
Comment (martinspetlik) [inline src/mlmc/quantity.py]: assert

File: src/mlmc/quantity.py (hunk 1 above)
Comment (martinspetlik) [reply src/mlmc/quantity.py:567]: #165

File: src/mlmc/quantity_types.py (hunk 7 above)
Comment (martinspetlik) [reply src/mlmc/quantity_types.py:76]: #164

**Pull request merged**